engine/lib/dependencies.ml

```diff
@@ -270,13 +286,79 @@ module Make (F : Features.T) = struct
     in
-    let lookup (name : concrete_ident) =
-      List.find ~f:(ident_of >> Concrete_ident.equal name) items
+    let stable_g =
+      let to_index =
+        items
+        |> List.mapi ~f:(fun i item -> (item.ident, i))
+        |> Map.of_alist_exn (module Concrete_ident)
+        |> Map.find
+      in
+      ItemGraph.Map_G_GInt.filter_map
+        (to_index *** to_index >> uncurry Option.both)
+        g
+    in
+    let stable_g =
+      List.foldi items ~init:stable_g ~f:(fun i g _ ->
+          ItemGraph.GInt.add_vertex g i)
+    in
+    let items' =
+      let items_array = Array.of_list items in
+      let lookup (index : int) = items_array.(index) in
+      ItemGraph.Topological.fold List.cons stable_g [] |> List.map ~f:lookup
     in
+    (* Stable topological sort doesn't guarantee to group cycles together.
+       We make this correction to ensure mutually recursive items are grouped. *)
     let items' =
-      ItemGraph.Topological.fold List.cons g []
-      |> List.filter_map ~f:lookup |> List.rev
+      let cycles =
+        ItemGraph.MutRec.SCC.scc_list g
+        |> List.filter ~f:(fun cycle -> List.length cycle > 1)
+      in
+      List.fold items' ~init:[] ~f:(fun acc item ->
+          match
+            List.find cycles ~f:(fun cycle ->
+                List.mem cycle item.ident ~equal:[%eq: concrete_ident])
+          with
+          | Some _
+            when List.exists acc ~f:(fun els ->
+                     List.mem els item ~equal:[%eq: item]) ->
+              [] :: acc
+          | Some cycle ->
+              List.map cycle ~f:(fun ident ->
+                  List.find_exn items ~f:(fun item ->
+                      [%eq: concrete_ident] item.ident ident))
+              :: acc
+          | None -> [ item ] :: acc)
+      |> List.concat
+    in
+    (* Quote items must be placed right before or after their origin *)
+    let items' =
+      let before_quotes, after_quotes, _ =
+        List.partition3_map items' ~f:(fun item ->
+            match item.v with
+            | Quote { origin; _ } -> (
+                match origin.position with
+                | `Before -> `Fst (origin, item)
+                | `After -> `Snd (origin, item)
+                | `Replace -> `Trd ())
+            | _ -> `Trd ())
+      in
+      let move_quote before origin quote_item =
+        List.concat_map ~f:(fun item ->
+            if [%eq: concrete_ident] origin.item_ident item.ident then
+              if before then [ quote_item; item ] else [ item; quote_item ]
+            else if [%eq: concrete_ident] quote_item.ident item.ident then []
+            else [ item ])
+      in
+      let before_quotes = List.rev before_quotes in
+      let items' =
+        List.fold before_quotes ~init:items'
+          ~f:(fun items' (origin, quote_item) ->
+            move_quote true origin quote_item items')
+      in
+      List.fold after_quotes ~init:items' ~f:(fun items' (origin, quote_item) ->
```

Comment:
Since we are always applying sorting before late skip, maybe this special treatment of replace is not useful.
But maybe we don't care